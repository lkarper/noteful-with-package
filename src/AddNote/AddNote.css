.add-note-form {
    display: flex;
    flex-direction: column;
}

.add-note-container {
    margin: 0 20px;
}

.add-note-form label {
    margin: 5px;
}

.add-note-form button {
    margin-top: 10px;
}

.note-content {
    width: 500px;
    height: 300px;
}