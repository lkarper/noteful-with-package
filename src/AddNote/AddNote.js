import React, { Component } from 'react';
import PropTypes from 'prop-types';
import NotesContext from '../NotesContext/NotesContext';
import ValidationError from '../ValidationError/ValidationError';
import './AddNote.css';

class AddNote extends Component {

    static contextType = NotesContext;

    static defaultProps = {
        location: {
            state: {
                folderId: '',
            },
        },
    };

    state = {
        name: {
            value: '',
            touched: false,
        },
        folderId: {
            value: '',
            touched: false,
        },
        content: {
            value: '',
            touched: false,
        },
        error: null,
    }

    updateName = (name) => {
        this.setState({
            name: {
                value: name,
                touched: true,
            },
        });
    }

    updateFolderId = (folderId) => {
        this.setState({
            folderId: {
                value: parseInt(folderId),
                touched: true,
            },
        });
    }

    updateContent = (content) => {
        this.setState({
            content: {
                value: content,
                touched: true,
            },
        });
    }

    handleAddNote = (event, cb) => {
        event.preventDefault();
        const newNote = {
            name: this.state.name.value,
            date_modified: (new Date()).toJSON(),
            folder: parseInt(this.state.folderId.value),
            content: this.state.content.value,
        };

        fetch('http://localhost:8000/api/notes', {
            method: 'POST',
            body: JSON.stringify(newNote),
            headers: {
                'content-type': 'application/json'
            }
        })
            .then(response => {
                if (response.ok) {
                    return response.json();
                }
                throw new Error(response.message);
            })
            .then(data => {
                cb(data);
                this.props.history.push(`/note/${data.id}`);
            })
            .catch(error => {
                console.log(error);
                this.setState({ error: error.message });
            });
    }

    validateNoteName = () => {
        const name = this.state.name.value.trim();
        if (name.length === 0) {
            return 'The note must be named and the name must contain at least one character.';
        }
    }

    validateFolderChoice = () => {
        if (!this.state.folderId.value) {
            return 'You must select a folder in which to place your note.';
        }
    }

    validateContent = () => {
        const content = this.state.content.value.trim();
        if (content.length === 0) {
            return 'You must add content to your note and it must contain at least one character.';
        }
    }

    componentDidMount() {
        if (this.props.location.state) {
            this.setState({
                folderId: {
                    value: parseInt(this.props.location.state.folderId),
                    touched: true,
                },
            })
        }
    }

    render() {

        const nameError = this.validateNoteName();
        const folderError = this.validateFolderChoice();
        const contentError = this.validateContent();

        const folderRadios = this.context.folders.map(folder => 
                <div key={folder.id}>
                    <input 
                        type="radio" 
                        id={folder.id} 
                        name="folder" 
                        value={folder.id} 
                        onChange={e => this.updateFolderId(e.currentTarget.value)} 
                        checked={this.state.folderId.value === folder.id}
                        aria-invalid={folderError ? true : false}
                        aria-describedby="folder-aria-required"
                        required 
                    />
                    <label htmlFor={folder.id}>{folder.name}</label>
                </div>
            );

        const { error } = this.state;
        const errorHTML = (
            <div className="folder-error">
                <h2>Looks like something went wrong: {error}.</h2>
                <p>A new note could not be created.  Please try again later.</p>
            </div>
        );

        return (
            <div className="add-note-container">
                <form 
                    className="add-note-form"
                    onSubmit={e => this.handleAddNote(e, this.context.addNote)}
                    >
                    <label htmlFor="add-note">Enter name for a new note: </label>
                    <input 
                        type="text" 
                        id="add-note" 
                        name="add-note" 
                        className="note-name" 
                        onChange={e => this.updateName(e.currentTarget.value)}
                        aria-required="true"
                        aria-invalid={nameError ? true : false} 
                        aria-describedby="note-aria-required"
                        required 
                    />
                    <div id="note-aria-required">
                        You must supply a name for your note.
                    </div>
                    {this.state.name.touched && <ValidationError message={nameError} />}
                    <fieldset>
                        <legend>Select a folder to put the new note in: </legend>
                        {folderRadios}
                    </fieldset>
                    <div id="folder-aria-required">
                        You must select a folder in which to place your note.
                    </div>
                    {this.state.folderId.touched && <ValidationError message={folderError} />}
                    <label htmlFor="content">Enter note content here:</label>
                    <textarea
                        id="content" 
                        name="content" 
                        rows="5" 
                        cols="75" 
                        className="note-content"
                        onChange={e => this.updateContent(e.currentTarget.value)} 
                        placeholder="Type your note here..." 
                        aria-required="true"
                        aria-invalid={contentError ? true : false}
                        aria-describedby="content-aria-required"
                        required>
                    </textarea>
                    <div id="content-aria-required">
                        You must add content to your note.
                    </div>
                    {this.state.content.touched && <ValidationError message={contentError} />}
                    <button 
                        type="submit"
                        disabled={nameError || folderError || contentError ? true : false}
                        >
                            Add Note
                    </button> 
                </form>
                {error ?  errorHTML : ''}
            </div>
        );
    }
}

AddNote.propTypes = {
    location: PropTypes.object.isRequired,
    history: PropTypes.object.isRequired,
}

export default AddNote;