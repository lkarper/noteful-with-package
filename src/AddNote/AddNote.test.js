import React from 'react';
import ReactDOM from 'react-dom';
import { mount } from 'enzyme';
import toJSON from 'enzyme-to-json';
import AddNote from './AddNote';

describe('AddNote component', () => {

    const context = {
        folders: [
            {
                name: "Test",
                id: 1,
            }
        ]
    }

    it('renders the UI without crashing', () => {
        const div = document.createElement('div');
        ReactDOM.render(<AddNote />, div);
        ReactDOM.unmountComponentAtNode(div);
    });

    it('renders the UI as expected', () => {
        const wrapper = mount(<AddNote />);
        expect(toJSON(wrapper)).toMatchSnapshot();
        wrapper.unmount();
    });

    it('renders an error message when the note name input contains no characters', () => {
        const wrapper = mount(<AddNote />, { context });
        wrapper.setContext({
            folders: [
                {
                    name: "Test",
                    id: 1,
                }
            ]
        });
        wrapper.find('.note-name').simulate('change', { currentTarget: {name: 'add-note', value: ' '} });
        expect(toJSON(wrapper)).toMatchSnapshot();
        wrapper.unmount();
    });

    it('renders an error message when no folder is clicked', () => {
        const wrapper = mount(<AddNote />, { context });
        wrapper.setContext({
            folders: [
                {
                    name: "Test",
                    id: 1,
                }
            ]
        });
        wrapper.find('.note-name').simulate('change', { currentTarget: {name: 'add-note', value: 'valid note name'} });
        wrapper.find('textarea').simulate('change', { currentTarget: {name: 'content', value: ' '} });
        expect(toJSON(wrapper)).toMatchSnapshot();
        wrapper.unmount();
    });

    it('does not display an error message when folder is clicked', () => {
        const wrapper = mount(<AddNote />, { context });
        wrapper.setContext({
            folders: [
                {
                    name: "Test",
                    id: 1,
                }
            ]
        });
        wrapper.setState({
            folderId: {
                value: 'b07162f0-ffaf-11e8-8eb2-f2801f1b9fd1',
                touched: true,
            },
        });
        expect(toJSON(wrapper)).toMatchSnapshot();
        wrapper.unmount();
    });

    it('renders an error message when no content is added to the note', () => {
        const wrapper = mount(<AddNote />, { context });
        wrapper.setContext({
            folders: [
                {
                    name: "Test",
                    id: 1,
                }
            ]
        });
        wrapper.find('.note-content').simulate('change', { currentTarget: {name: 'content', value: ' '} });
        expect(toJSON(wrapper)).toMatchSnapshot();
        wrapper.unmount();
    })

    it('disables submit button when input is blank', () => {
        const wrapper = mount(<AddNote />, { context });
        wrapper.setContext({
            folders: [
                {
                    name: "Test",
                    id: 1,
                }
            ]
        });
        wrapper.find('button').simulate('click');
        expect(toJSON(wrapper)).toMatchSnapshot();
        wrapper.unmount();
    });

    it('enables submit button when valid note name, folder, and content are provided', () => {
        const wrapper = mount(<AddNote />, { context });
        wrapper.setContext({
            folders: [
                {
                    name: "Test",
                    id: 1,
                }
            ]
        });
        wrapper.setState({
                name: {
                    value: 'name',
                    touched: true,
                },
                folderId: {
                    value: 'b07162f0-ffaf-11e8-8eb2-f2801f1b9fd1',
                    touched: true,
                },
                content: {
                    value: 'content',
                    touched: true,
                },
                error: null,
            });
        expect(toJSON(wrapper)).toMatchSnapshot();
        wrapper.unmount();
    });

});