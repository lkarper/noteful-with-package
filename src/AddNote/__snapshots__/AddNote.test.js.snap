// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddNote component disables submit button when input is blank 1`] = `
<AddNote
  location={
    Object {
      "state": Object {
        "folderId": "",
      },
    }
  }
>
  <div
    className="add-note-container"
  >
    <form
      className="add-note-form"
      onSubmit={[Function]}
    >
      <label
        htmlFor="add-note"
      >
        Enter name for a new note: 
      </label>
      <input
        className="note-name"
        id="add-note"
        name="add-note"
        onChange={[Function]}
        required={true}
        type="text"
      />
      <fieldset>
        <legend>
          Select a folder to put the new note in: 
        </legend>
      </fieldset>
      <ValidationError
        message="You must select a folder in which to place your note."
      >
        <div
          className="validation-error"
        >
          You must select a folder in which to place your note.
        </div>
      </ValidationError>
      <label
        htmlFor="content"
      >
        Enter note content here:
      </label>
      <textarea
        className="note-content"
        cols="75"
        id="content"
        name="content"
        onChange={[Function]}
        placeholder="Type your note here..."
        required={true}
        rows="5"
      />
      <button
        disabled={true}
        type="submit"
      >
        Add Note
      </button>
    </form>
  </div>
</AddNote>
`;

exports[`AddNote component does not display an error message when folder is clicked 1`] = `
<AddNote
  history={
    Object {
      "push": [Function],
    }
  }
  location={
    Object {
      "state": Object {
        "folderId": "",
      },
    }
  }
>
  <div
    className="add-note-container"
  >
    <h2>
      Loading...
    </h2>
  </div>
</AddNote>
`;

exports[`AddNote component enables submit button when valid note name, folder, and content are provided 1`] = `
<AddNote
  history={
    Object {
      "push": [Function],
    }
  }
  location={
    Object {
      "state": Object {
        "folderId": "",
      },
    }
  }
>
  <div
    className="add-note-container"
  >
    <h2>
      Loading...
    </h2>
  </div>
</AddNote>
`;

exports[`AddNote component renders an error message when no content is added to the note 1`] = `
<AddNote
  location={
    Object {
      "state": Object {
        "folderId": "",
      },
    }
  }
>
  <div
    className="add-note-container"
  >
    <form
      className="add-note-form"
      onSubmit={[Function]}
    >
      <label
        htmlFor="add-note"
      >
        Enter name for a new note: 
      </label>
      <input
        className="note-name"
        id="add-note"
        name="add-note"
        onChange={[Function]}
        required={true}
        type="text"
      />
      <fieldset>
        <legend>
          Select a folder to put the new note in: 
        </legend>
      </fieldset>
      <ValidationError
        message="You must select a folder in which to place your note."
      >
        <div
          className="validation-error"
        >
          You must select a folder in which to place your note.
        </div>
      </ValidationError>
      <label
        htmlFor="content"
      >
        Enter note content here:
      </label>
      <textarea
        className="note-content"
        cols="75"
        id="content"
        name="content"
        onChange={[Function]}
        placeholder="Type your note here..."
        required={true}
        rows="5"
      />
      <ValidationError
        message="You must add content to your note and it must contain at least one character."
      >
        <div
          className="validation-error"
        >
          You must add content to your note and it must contain at least one character.
        </div>
      </ValidationError>
      <button
        disabled={true}
        type="submit"
      >
        Add Note
      </button>
    </form>
  </div>
</AddNote>
`;

exports[`AddNote component renders an error message when no folder is clicked 1`] = `
<AddNote
  location={
    Object {
      "state": Object {
        "folderId": "",
      },
    }
  }
>
  <div
    className="add-note-container"
  >
    <form
      className="add-note-form"
      onSubmit={[Function]}
    >
      <label
        htmlFor="add-note"
      >
        Enter name for a new note: 
      </label>
      <input
        className="note-name"
        id="add-note"
        name="add-note"
        onChange={[Function]}
        required={true}
        type="text"
      />
      <ValidationError
        message="The note must be named and the name must contain at least one character."
      >
        <div
          className="validation-error"
        >
          The note must be named and the name must contain at least one character.
        </div>
      </ValidationError>
      <fieldset>
        <legend>
          Select a folder to put the new note in: 
        </legend>
      </fieldset>
      <ValidationError
        message="You must select a folder in which to place your note."
      >
        <div
          className="validation-error"
        >
          You must select a folder in which to place your note.
        </div>
      </ValidationError>
      <label
        htmlFor="content"
      >
        Enter note content here:
      </label>
      <textarea
        className="note-content"
        cols="75"
        id="content"
        name="content"
        onChange={[Function]}
        placeholder="Type your note here..."
        required={true}
        rows="5"
      />
      <ValidationError
        message="You must add content to your note and it must contain at least one character."
      >
        <div
          className="validation-error"
        >
          You must add content to your note and it must contain at least one character.
        </div>
      </ValidationError>
      <button
        disabled={true}
        type="submit"
      >
        Add Note
      </button>
    </form>
  </div>
</AddNote>
`;

exports[`AddNote component renders an error message when the note name input contains no characters 1`] = `
<AddNote
  location={
    Object {
      "state": Object {
        "folderId": "",
      },
    }
  }
>
  <div
    className="add-note-container"
  >
    <form
      className="add-note-form"
      onSubmit={[Function]}
    >
      <label
        htmlFor="add-note"
      >
        Enter name for a new note: 
      </label>
      <input
        className="note-name"
        id="add-note"
        name="add-note"
        onChange={[Function]}
        required={true}
        type="text"
      />
      <ValidationError
        message="The note must be named and the name must contain at least one character."
      >
        <div
          className="validation-error"
        >
          The note must be named and the name must contain at least one character.
        </div>
      </ValidationError>
      <fieldset>
        <legend>
          Select a folder to put the new note in: 
        </legend>
      </fieldset>
      <ValidationError
        message="You must select a folder in which to place your note."
      >
        <div
          className="validation-error"
        >
          You must select a folder in which to place your note.
        </div>
      </ValidationError>
      <label
        htmlFor="content"
      >
        Enter note content here:
      </label>
      <textarea
        className="note-content"
        cols="75"
        id="content"
        name="content"
        onChange={[Function]}
        placeholder="Type your note here..."
        required={true}
        rows="5"
      />
      <button
        disabled={true}
        type="submit"
      >
        Add Note
      </button>
    </form>
  </div>
</AddNote>
`;

exports[`AddNote component renders the UI as expected 1`] = `
<AddNote
  history={
    Object {
      "push": [Function],
    }
  }
  location={
    Object {
      "state": Object {
        "folderId": "",
      },
    }
  }
>
  <div
    className="add-note-container"
  >
    <h2>
      Loading...
    </h2>
  </div>
</AddNote>
`;
